muikku-core-plugins/src/main/resources/META-INF/resources/scripts/src/components/credentials/body/return-credentials.tsx

```diff
@@ -0,0 +1,164 @@
+import * as React from 'react';
+import { connect, Dispatch} from 'react-redux';
+import mApi, { MApiError } from '~/lib/mApi';
+import promisify from '~/util/promisify';
+import Panel from '~/components/general/panel';
+import Button from '~/components/general/button';
+import "~/sass/elements/form-elements.scss";
+import "~/sass/elements/form.scss";
+import { displayNotification, DisplayNotificationTriggerType } from '~/actions/base/notifications';
+import {i18nType} from '~/reducers/base/i18n';
+import {StateType} from '~/reducers';
+import { bindActionCreators } from 'redux';
+import LoginButton from '../../base/login-button';
+
+interface ReturnCredentialsProps {
+  secret: string,
+  displayNotification: DisplayNotificationTriggerType,
+  i18n: i18nType,
+}
+
+interface ReturnCredentialsState {
+  username: string,
+  newPassword: string,
+  newPasswordConfirm: string,
+  locked: boolean,
+  handled: boolean
+}
+
+class ReturnCredentials extends React.Component<ReturnCredentialsProps, ReturnCredentialsState> {
+  constructor(props: ReturnCredentialsProps){
+    super(props);
+    this.state = {
+       username: "",
+       newPassword: "",
+       newPasswordConfirm:"",
+       locked: false,
+       handled: false
+    }
+  }
+  async componentDidMount() {
+    try {
+    const data:any = await (promisify(mApi().forgotpassword.credentialReset.read(this.props.secret), 'callback')());
+      if (data.username != null) {
+        this.setState({username: data.username});
+      } 
+
+    } catch (err){
+      this.props.displayNotification(this.props.i18n.text.get("plugin.forgotpassword.changeCredentials.messages.error.hashLoadFailed", err.message), "error");
+      if (!(err instanceof MApiError)){
+        this.props.displayNotification(err.message, "error");
+      }
+
+    }
+  }
+  
+  handleNewCredentials() {
+    let newPassword1 = this.state.newPassword;
+    let newPassword2 = this.state.newPasswordConfirm;
+    let userName = this.state.username;
+    
+    if(userName == "") {
+      this.props.displayNotification(this.props.i18n.text.get("plugin.forgotpassword.changeCredentials.messages.error.empty.username"), "error");
+      return;
+    }
+    
+    if (newPassword1 !== newPassword2) {
+      this.props.displayNotification(this.props.i18n.text.get("plugin.forgotpassword.changeCredentials.messages.error.passwordsDontMatch"), "error");
+      return;
+    }
+    
+    if (newPassword1 == "" || newPassword2 == "") {
+      this.props.displayNotification(this.props.i18n.text.get("plugin.forgotpassword.changeCredentials.messages.error.empty.passwords"), "error");
+      return;
+    }
+    
+    this.setState({
+      locked: true
+    });
+    
+    let values = {
+        secret : this.props.secret,
+        password : newPassword1,
+        username : userName
+    };
+    
+     mApi().forgotpassword.credentialReset.create(values).callback((err: any, result: any)=>{
```

Comment:
We need an action that should be credetialReset and supports these callbacks.

Only in exceptional situations we call mApi from the components.

You can add it in /actions/credentials/index.ts even if it's only one action

Reply:
I thought this was the exceptional situation. I mean a whole application store because of one function?